lang/cs/Org.Apache.REEF.Network/Elastic/Task/CommunicationLayer.cs

```diff
@@ -0,0 +1,304 @@
+        internal void Send(
+            string destination,
+            ElasticGroupCommunicationMessage message,
+            CancellationTokenSource cancellationSource)
+        {
+            if (message == null)
+            {
+                throw new ArgumentNullException(nameof(message));
+            }
+            if (string.IsNullOrEmpty(destination))
+            {
+                throw new ArgumentNullException(nameof(destination));
+            }
+            if (_disposed)
+            {
+                Log.Log(Level.Warning, "Received send message request after disposing: Ignoring.");
+                return;
+            }
+
+            IIdentifier destId = _idFactory.Create(destination);
+            int retry = 0;
+
+            while (!Send(destId, message))
+            {
+                if (retry > _retrySending)
+                {
+                    throw new IllegalStateException($"Unable to send message after retying {retry} times.");
+                }
+                Thread.Sleep(_timeout);
+
+                retry++;
+            }
+        }
+
+        /// <summary>
+        /// Forward the received message to the target <see cref="IOperatorTopologyWithCommunication"/>.
+        /// </summary>
+        /// <param name="remoteMessage">The received message</param>
+        public abstract void OnNext(IRemoteMessage<NsMessage<ElasticGroupCommunicationMessage>> remoteMessage);
+
+        /// <summary>
+        /// Checks if the identifier is registered with the name server.
+        /// Throws exception if the operation fails more than the retry count.
+        /// </summary>
+        /// <param name="identifiers">The identifier to look up</param>
+        /// <param name="cancellationSource">The token to cancel the operation</param>
+        /// <param name="removed">Nodes that got removed during task registration</param>
+        public void WaitForTaskRegistration(
+            IList<string> identifiers,
+            CancellationTokenSource cancellationSource,
+            ConcurrentDictionary<string, byte> removed = null)
+        {
+            if (removed == null)
+            {
+                removed = new ConcurrentDictionary<string, byte>();
+            }
+
+            IList<string> foundList = new List<string>();
+
+            for (var i = 0; i < _retryRegistration; i++)
+            {
+                if (cancellationSource != null && cancellationSource.Token.IsCancellationRequested)
+                {
+                    Log.Log(Level.Warning, "WaitForTaskRegistration is canceled in retryCount {0}.", i);
+                    throw new OperationCanceledException("WaitForTaskRegistration is canceled");
+                }
+
+                Log.Log(Level.Info, "WaitForTaskRegistration, in retryCount {0}.", i);
+                foreach (var identifier in identifiers)
+                {
+                    var notFound = !foundList.Contains(identifier);
```

Comment:
`ISet.Add()` returns `false` if the element has already been there, so we can rewrite the condition as
```C#
if (!foundList.Add(identifier))
{
    if (removed.ContainsKey(identifier))
    {
        Log.Log(Level.Verbose,
            "WaitForTaskRegistration, dependent id {0} was removed at loop {1}.", identifier, i);
    }
    else if (Lookup(identifier))
    {
        Log.Log(Level.Verbose,
            "WaitForTaskRegistration, find a dependent id {0} at loop {1}.", identifier, i);
    }
}
```

Reply:
Not sure that this works: in the case the node is not in found list and was neither removed (`removed.ContainsKey`) nor reachable (`Lookup`) we are adding it anyway to the found list which is not correct.